### PR TITLE
p2p: extend e2e tests for new p2p framework

### PR DESCRIPTION
**Pull request by tychoish**
This is mostly just an experiment in seeing if wiring this up will
even work.

It might make sense to back out of this a bit and just keep the
parameterization without enabling it as an axis in the tests, for now.

### CODE REVIEW

File: p2p/peermanager.go [hunk 1]
```diff
@@ -257,11 +257,6 @@ func (o *PeerManagerOptions) optimize() {
 //   lower-scored to evict.
 // - EvictNext: pick peer from evict, mark as evicting.
 // - Disconnected: unmark connected, upgrading[from]=to, evict, evicting.
-//
-// FIXME: The old stack supports ABCI-based peer ID filtering via
-// /p2p/filter/id/<ID> queries, we should implement this here as well by taking
-// a peer ID filtering callback in PeerManagerOptions and configuring it during
-// Node setup.
```
Comment (cmwaters) [inline p2p/peermanager.go:264]: Has this actually been solved

File: node/node.go [hunk 2]
```diff
@@ -1321,12 +1316,12 @@ func (n *Node) OnStart() error {
 
 	n.isListening = true
 
-	n.Logger.Info("p2p service", "legacy_enabled", useLegacyP2P)
+	n.Logger.Info("p2p service", "legacy_enabled", n.config.P2P.UseNewP2P)
```
Comment (cmwaters) [inline node/node.go:1319]:
```suggestion
	n.Logger.Info("p2p service", "legacy_enabled", !n.config.P2P.UseNewP2P)
```

**Review by cmwaters — COMMENTED** [review]
The major sort of change I see here is to change the boolean variable that dictates which P2P version we use from being an environment variable to a config variable. 

What are the sort of tradeoffs with this? I would imagine that using a config variable is probably better because it makes the scope local to that node right

File: p2p/peermanager.go (hunk 1 above)
Comment (tychoish) [reply p2p/peermanager.go:264]: yes!

File: test/e2e/generator/generate.go [hunk 3]
```diff
@@ -16,6 +16,7 @@ var (
 	testnetCombinations = map[string][]interface{}{
 		"topology":      {"single", "quad", "large"},
 		"ipv6":          {false, true},
+		"useNewP2P":     {false, true, 2, 3},
```
Comment (cmwaters) [inline test/e2e/generator/generate.go:19]: What's with the `2, 3` part?

Comment (tychoish) [reply test/e2e/generator/generate.go:19]: for these networks, make a legacy network but have half or a third of the nodes use the new stack.

Comment (cmwaters) [reply test/e2e/generator/generate.go:19]: also any field in this map gets systematically tested of all permutations so you'll need to be careful about how many you add. I think prior we generated around 90 testnets for the nightlies. With this addition we will have around 360 testnets

Comment (cmwaters) [reply test/e2e/generator/generate.go:19]: > What's with the `2, 3` part? 

Ah this is to account for mixed testnets

Comment (tychoish) [reply test/e2e/generator/generate.go:19]: cool, I think it makes sense to test "on, off, half" suites, but not "third", which cuts back some? 

How long does a single testnet run take?

Comment (cmwaters) [reply test/e2e/generator/generate.go:19]: I think I would just have three combinations: "new", "old", "split"

File: test/e2e/generator/generate.go [hunk 4]
```diff
@@ -134,8 +150,13 @@ func generateTestnet(r *rand.Rand, opt map[string]interface{}) (e2e.Manifest, er
...
 		}
-		manifest.Nodes[fmt.Sprintf("full%02d", i)] = generateNode(
-			r, e2e.ModeFull, startAt, manifest.InitialHeight, false)
+		node := generateNode(r, e2e.ModeFull, startAt, manifest.InitialHeight, false)
+		if p2pNodeFactor == 0 {
+			node.UseNewP2P = manifest.UseNewP2P
+		} else if p2pNodeFactor%i == 0 {
+			node.UseNewP2P = !manifest.UseNewP2P
+		}
+		manifest.Nodes[fmt.Sprintf("full%02d", i)] = node
```
Comment (cmwaters) [inline test/e2e/generator/generate.go:159]: I can't actually point to the correct piece of code but you've missed implementing this when we generate the validators

File: test/e2e/generator/generate.go (hunk 3 above)
Comment (cmwaters) [reply test/e2e/generator/generate.go]: > How long does a single testnet run take?

it varies quite a bit depending on the size and the amount of perturbations. I would say between 3 - 20 seconds

Comment (tychoish) [reply test/e2e/generator/generate.go]: I think we 're on the same page.

**Review by cmwaters — APPROVED** [review]
I ran a couple of generated tests with the new framework and it all seemed to work smoothly (perhaps surprisingly :) )

**Comment by tychoish** [comment]
It turns out that I had failed to update the config template to include the new option, so our tests worked because they were all (actually) using the old p2p stack! 

I've fixed things, along with a collection of issues with the metrics collection that I think were leading to some runtime panics. Now I'm seeing tests timeout, which both gives me more confidence that this change is actually exercising the code that we want it to (yay), but means that we shouldn't merge it quite yet, without a bit more work/validation/etc. (It's possible that my local test setup is also not trustworthy.)

File: config/config.go [hunk 5]
```diff
@@ -607,6 +607,15 @@ type P2PConfig struct { //nolint: maligned
 	// Testing params.
 	// Force dial to fail
 	TestDialFail bool `mapstructure:"test-dial-fail"`
+
+	// Mostly for testing, use rather than environment variables
+	// to turn on the new P2P stack.
+	UseNewP2P bool `mapstructure:"use-new-p2p"`
```
Comment (alexanderbez) [inline config/config.go:613]: nit: this should've been called `enable-legacy` or just `legacy` or something.

**Pull request merged**